src/wat-writer.cc

```diff
@@ -807,9 +871,46 @@ void WatWriter::PushExpr(const Expr* expr,
     expr_tree_stack_.emplace_back(expr);
     FlushExprTreeStack();
   }
+  if (s_trace) {
+    fprintf(stderr, "-> PushExpr(%s, %" PRIindex ", %" PRIindex")\n",
+            GetExprTypeName(*expr), operand_count, result_count);
+  }
+}
+
+void WatWriter::PushCatch(const Catch* catch_) {
+  if (s_trace)
+    fprintf(stderr, "-> PushCatch()\n");
+  ExprTree tree(catch_);
+  expr_tree_stack_.push_back(std::move(tree));
+  FlushExprTreeStack();
+  if (s_trace)
+    fprintf(stderr, "<- PushCatch()\n");
+}
+
+void WatWriter::FlushCatchTree(const ExprTree& expr_tree) {
```

Comment:
I don't understand why catches need to be pushed on the stack. It seems like all of this code could be inlined directly into `FlushExprTree` below since you always immediately flush the stack after you push a new catch on.